From: Oleg Nesterov <oleg@redhat.com>
Subject: do_io_accounting: use sig->stats_lock
Date: Mon, 23 Oct 2023 17:34:05 +0200

Rather than lock_task_sighand(), sig->stats_lock was specifically designed
for this type of use.

This way the "if (whole)" branch runs lockless in the likely case.

Link: https://lkml.kernel.org/r/20231023153405.GA4639@redhat.com
Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Cc: "Eric W. Biederman" <ebiederm@xmission.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 fs/proc/base.c |   20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

--- a/fs/proc/base.c~do_io_accounting-use-sig-stats_lock
+++ a/fs/proc/base.c
@@ -2973,7 +2973,6 @@ static const struct file_operations proc
 static int do_io_accounting(struct task_struct *task, struct seq_file *m, int whole)
 {
 	struct task_io_accounting acct;
-	unsigned long flags;
 	int result;
 
 	result = down_read_killable(&task->signal->exec_update_lock);
@@ -2985,15 +2984,24 @@ static int do_io_accounting(struct task_
 		goto out_unlock;
 	}
 
-	if (whole && lock_task_sighand(task, &flags)) {
+	if (whole) {
 		struct signal_struct *sig = task->signal;
 		struct task_struct *t;
+		unsigned int seq = 1;
+		unsigned long flags;
 
-		acct = sig->ioac;
-		__for_each_thread(sig, t)
-			task_io_accounting_add(&acct, &t->ioac);
+		rcu_read_lock();
+		do {
+			seq++; /* 2 on the 1st/lockless path, otherwise odd */
+			flags = read_seqbegin_or_lock_irqsave(&sig->stats_lock, &seq);
 
-		unlock_task_sighand(task, &flags);
+			acct = sig->ioac;
+			__for_each_thread(sig, t)
+				task_io_accounting_add(&acct, &t->ioac);
+
+		} while (need_seqretry(&sig->stats_lock, seq));
+		done_seqretry_irqrestore(&sig->stats_lock, seq, flags);
+		rcu_read_unlock();
 	} else {
 		acct = task->ioac;
 	}
_
